msgraph/authentication_methods_test.go

```diff
@@ -0,0 +1,391 @@
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/manicminer/hamilton/auth"
+	"github.com/manicminer/hamilton/internal/test"
+	"github.com/manicminer/hamilton/internal/utils"
+	"github.com/manicminer/hamilton/msgraph"
+	"github.com/manicminer/hamilton/odata"
+)
+
+type AuthenticationMethodsClientTest struct {
+	connection   *test.Connection
+	client       *msgraph.AuthenticationMethodsClient
+	randomString string
+}
+
+func TestAuthenticationMethodsClient(t *testing.T) {
+	rs := test.RandomString()
+	c := AuthenticationMethodsClientTest{
+		connection:   test.NewConnection(auth.MsGraph, auth.TokenVersion2),
+		randomString: rs,
+	}
+
+	c.client = msgraph.NewAuthenticationMethodsClient(c.connection.AuthConfig.TenantID)
+	c.client.BaseClient.Authorizer = c.connection.Authorizer
+
+	u := UsersClientTest{
+		connection:   test.NewConnection(auth.MsGraph, auth.TokenVersion2),
+		randomString: rs,
+	}
+
+	u.client = msgraph.NewUsersClient(u.connection.AuthConfig.TenantID)
+	u.client.BaseClient.Authorizer = u.connection.Authorizer
+
+	user := testUsersClient_Create(t, u, msgraph.User{
+		AccountEnabled:    utils.BoolPtr(true),
+		DisplayName:       utils.StringPtr("test-user-authenticationmethods"),
+		MailNickname:      utils.StringPtr(fmt.Sprintf("test-user-authenticationmethods-%s", c.randomString)),
+		UserPrincipalName: utils.StringPtr(fmt.Sprintf("test-user-authenticationmethods-%s@%s", c.randomString, c.connection.DomainName)),
+		PasswordProfile: &msgraph.UserPasswordProfile{
+			Password: utils.StringPtr(fmt.Sprintf("IrPa55w0rd%s", c.randomString)),
+		},
+	})
+
+	_ = testAuthMethods_List(t, c, *user.ID)
+	_ = testAuthMethods_ListFido2Methods(t, c, *user.ID)
+	_ = testAuthMethods_ListMicrosoftAuthenticatorMethods(t, c, *user.ID)
+	_ = testAuthMethods_ListWindowsHelloMethods(t, c, *user.ID)
+	tempAccessPass := testAuthMethods_CreateTemporaryAccessPassMethod(t, c, *user.ID)
+	_ = testAuthMethods_GetTemporaryAccessPassMethod(t, c, *user.ID, *tempAccessPass.ID)
+	_ = testAuthMethods_ListTemporaryAccessPassMethods(t, c, *user.ID)
+	testAuthMethods_DeleteTemporaryAccessPassMethod(t, c, *user.ID, *tempAccessPass.ID)
+	phoneAuthMethod := testAuthMethods_CreatePhoneMethod(t, c, *user.ID)
+	_ = testAuthMethods_GetPhoneMethod(t, c, *user.ID, *phoneAuthMethod.ID)
+	_ = testAuthMethods_ListPhoneMethods(t, c, *user.ID)
+	phoneAuthMethod.PhoneNumber = utils.StringPtr("+44 07940123966")
+	testAuthMethods_UpdatePhoneMethod(t, c, *user.ID, *phoneAuthMethod)
+	testAuthMethods_EnablePhoneSMS(t, c, *user.ID, *phoneAuthMethod.ID)
+	testAuthMethods_DisablePhoneSMS(t, c, *user.ID, *phoneAuthMethod.ID)
+	testAuthMethods_DeletePhoneMethod(t, c, *user.ID, *phoneAuthMethod.ID)
+	emailAuthMethod := testAuthMethods_CreateEmailMethod(t, c, *user.ID)
+	_ = testAuthMethods_GetEmailMethod(t, c, *user.ID, *emailAuthMethod.ID)
+	_ = testAuthMethods_ListEmailMethods(t, c, *user.ID)
+	emailAuthMethod.EmailAddress = utils.StringPtr("test-user-authenticationmethods@updateddomain.com")
```

Comment:
@MarkDordoy Would prefer to use contoso.com or similar well-contained domain - other domains could feasibly be registered and potentially [ab]used

Reply:
Pushed :)